src/query/storages/fuse/fuse/src/operations/mutation/compact/merge_segments_transform.rs

```diff
@@ -0,0 +1,186 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::any::Any;
+use std::sync::Arc;
+
+use common_datablocks::DataBlock;
+use common_exception::ErrorCode;
+use common_exception::Result;
+use common_storages_table_meta::meta::Location;
+use common_storages_table_meta::meta::SegmentInfo;
+use common_storages_table_meta::meta::Statistics;
+use common_storages_table_meta::meta::Versioned;
+use itertools::Itertools;
+
+use crate::operations::mutation::compact::CompactSinkMeta;
+use crate::operations::mutation::AbortOperation;
+use crate::operations::mutation::BlockCompactMutator;
+use crate::operations::mutation::MutationMeta;
+use crate::pipelines::processors::port::InputPort;
+use crate::pipelines::processors::port::OutputPort;
+use crate::pipelines::processors::processor::Event;
+use crate::pipelines::processors::processor::ProcessorPtr;
+use crate::pipelines::processors::Processor;
+use crate::statistics::reducers::merge_statistics_mut;
+
+enum State {
+    Consume,
+    Merge(DataBlock),
+    Output,
+}
+
+pub struct MergeSegmentsTransform {
+    state: State,
+    inputs: Vec<Arc<InputPort>>,
+    output: Arc<OutputPort>,
+    cur_input_index: usize,
+    output_data: Option<DataBlock>,
+
+    // The order of the merged segments in snapshot.
+    pub merged_indices: Vec<usize>,
+    // locations all the merged segments.
+    pub merged_segments: Vec<Location>,
+    // summarised statistics of all the merged segments
+    pub merged_statistics: Statistics,
+    abort_operation: AbortOperation,
+}
+
+impl MergeSegmentsTransform {
+    pub fn try_create(
+        mutator: BlockCompactMutator,
+        inputs: Vec<Arc<InputPort>>,
+        output: Arc<OutputPort>,
+    ) -> Result<ProcessorPtr> {
+        Ok(ProcessorPtr::create(Box::new(MergeSegmentsTransform {
+            state: State::Consume,
+            inputs,
+            output,
+            cur_input_index: 0,
+            output_data: None,
+            merged_indices: mutator.unchanged_segment_indices,
+            merged_segments: mutator.unchanged_segment_locations,
+            merged_statistics: mutator.unchanged_segment_statistics,
+            abort_operation: AbortOperation::default(),
+        })))
+    }
+
+    fn get_current_input(&mut self) -> Option<Arc<InputPort>> {
+        let mut finished = true;
+        let mut index = self.cur_input_index;
+
+        loop {
+            let input = &self.inputs[index];
```

Comment:
Do we need to check the inputs' length?

Reply:
https://github.com/datafuselabs/databend/pull/8824/files/7700c95cbb534196d9743b80b314653afc4a682d#diff-9ea44ca0145e52317af2fa0e3f0c5e8a0018712dfd449a4ed10e82c79b6dd406R96-R98